Inject oAuthToken and bundleId

momo-ozawa opened the pull request:
## Description
This PR injects the bearer token and the bundle id from the app into MobilePayKIt.

## Changes
- Added a `MobilePayKitConfiguration` struct that holds the bearer token and the bundle id
- Renamed `PaymentManager` to `MobilePayKit`
- Added `MobilePayKitTests`

## How to test

### Running tests
1. Run MobilePayKitTests test target
2. ✅ Tests should pass

### Running the app

1. Configure the bearer token in `ProductListViewModel`
2. Run the iOS app
3. Purchase the product by tapping on it
4. Tap Confirm
5. Tap OK
6. On Xcode debugger menu, click on the Manage StoreKit Transactions menu item 
7. ✅ There should be a record of a product being purchased (See example below)

<img width="747" alt="Screen Shot 2021-07-19 at 16 32 32" src="https://user-images.githubusercontent.com/6711616/126186582-3f76559e-7fe5-45ae-9068-2754883f6351.png">

Commit:
Refactor: rename PaymentManger to MobilePayKit

Commit:
Add: MobilePayKitConfiguration struct

We need to add a memberwise initializer since we'll be initializing MobilePayKitConfiguration outside the MobilePayKit:

https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html#//apple_ref/doc/uid/TP40014097-CH41-ID3

"... if you want a public structure type to be initializable with a memberwise initializer when used in another module, you must provide a public memberwise initializer yourself as part of the type’s definition."

Commit:
Refactor: inject token and bundle id

Commit:
Add: MobilePayKitConfiguration test fixture

Commit:
Fix: inject config fixture

Commit:
Refactor: define TestPaymentTransaction fixture in a new file

leandroalonso commented on Source/MobilePayKit.swift:
```diff
@@ -1,3 +1,75 @@
+
+public class MobilePayKit: NSObject {
+
+    private static let sharedInstance = MobilePayKit()
```
I'm a little bit confused, we're applying the singleton pattern here but all the methods are static/class and they can be called directly.

I'd prefer to not have `class`/`static` method, since this will guarantee that if the object was created, a configuration was given, and you won't need to check for its existence.

It will also force people to inject this in order to use it, which I really like. :P

emilylaguna replied:
I agree with @leandroalonso.

momo-ozawa replied:
So my thinking was this:

1. I wanted to create a `sharedInstance`, because there should only be one instance of `MobilePayKit`
2. I wanted to inject the configs, but keep the init method private
3. I wanted to make it easier to call the MobilePayKit methods, so I made wrapper class methods which ultimately call the methods on the `sharedInstance`

To address point 2, I added a class method `configure`, which MUST be called before calling any other MobilePayKit methods. Otherwise, you get a fatalError. I could add an assert statement to make this more clear.

```
    public class func fetchProducts(completion: @escaping FetchCompletionCallback) {
        assert(configuration != nil, "MobilePayKit.configure(configuration:) must be called before calling this method")
        sharedInstance.fetchProducts(completion: completion)
    }
```

About point 3 --  I thought it would be nicer to be able to call `MobilePayKit.method` as opposed to `MobilePayKit.sharedInstance.method`, and the wrapper class methods are calling the instance methods anyway under the hood.

I'm more than happy to inject the config via init, I just wanted to clarify the approach I took. 

Let me know your thoughts! 😄

leandroalonso replied:
Oh, I see.

In this case, I think you can rely just on the `sharedInstance` and let the user to deal with how they want to use it. Take for example `UserDefaults.standard`, in a lot of places we assign this to a var to make it easier to use, instead of calling the `default` over and over.

The static methods favor inline use, which in turn makes testing almost impossible. Getting rid of them would probably solve the issues and prevent that. :)

nikola-milicevic replied:
@momo-ozawa I do understand the intent, and I like it, but I agree with @leandroalonso about mentioned shortcomings, so I vote for direct use of `shared` (not `sharedInstance` 😁) and injected dependencies.

momo-ozawa replied:
Thanks for the feedback y'all! Will update this PR shortly.

leandroalonso commented on Source/Services/AppStoreService.swift:
```diff
@@ -7,7 +7,7 @@ public typealias PurchaseCompletionCallback = (SKPaymentTransaction?) -> Void
 class AppStoreService: NSObject {
 
-    private let iapService: InAppPurchasesService
+    private let iapService: InAppPurchasesService?
```
Given in the constructor we assign this property a default value in case `nil` was given, there's no reason for this to be `Optional`.

momo-ozawa replied:
Fixed! 57c03e8

Commit:
Fix: make InAppPurchases non-optional

Commit:
Refactor: remove class methods and inject via init

Commit:
Refactor: make AppStoreService public

AppStoreService needs to have a public access control because we are passing it as a parameter in the MobilePayKit constructor.

Commit:
Add: configure method for MobilePayKit

Commit:
Refactor: call MobilePayKit methods on shared instance

Commit:
Add: AppStoreServiceProtocol

Commit:
Refactor: use protocol as a type

Commit:
Add: MockAppStoreService

Commit:
Add: MobilePayKitTests
